enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,543 @@
+
+| Field | Description |
+| ----- | ----------- |
+| Type | Type of condition in CamelCase or in foo.example.com/CamelCase. |
+| Status | Status of the condition, one of True, False, Unknown. |
+| ObservedGeneration | If set, this represents the .metadata.generation that the condition was set based upon. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
+| Reason | The reason for the condition's last transition in CamelCase. |
+| Message | A human readable message indicating details about the transition. |
+
+OLM will define a set of "OLM Supported Conditions" which will influence the behavior of OLM when present. The list of "OLM Supported Conditions" can be seen in the table below:
+
+| Condition Type | Effect |
+| -------------- | ------ |
+| Upgradeable | When `false`, OLM will not upgrade the operator.|
+| Important | OLM will highlight this condition to the Cluster Admin.|
+
+Operators will communicate their conditions through one of two methods:
+
+1. By setting conditions on CustomResource that the operator owns.
+2. By setting conditions on a `Probe` CustomResource directly.
+
+### Why allow operators to communicate status through CustomResources
+
+There are many reasons why operators should communicate their conditions through owned CustomResources:
+
+- CustomResources are the main entry point to an operator and operators are already expected to update the status of the CustomResource to communicate "Actual State".
+- By writing the condition to the status of an owned CustomResource, it becomes very clear which CR is responsible for changing OLM behavior.
+- Introducing an OLM owned CRD forces operators to integrate with OLM at a much deeper level than writing to a resource that they already manage. Operators would be required to detect if it is being managed by OLM and begin writing to the API. By writing to an owned CR, the Operator does not require OLM specific code and can be run independently of OLM.
+
+Allowing operators to write to the status of any CR creates a potential scaling issue if enough CRs are on cluster. OLM will avoid these performance issues by introducing the Status Probe Operator (SPO) which will reconcile the `Probe` resource. If an operator chooses to write to their own CRs, SPO will create a pod that will keep the status of the `Probe` resource synced with the conditions in the CRs.
+
+>Note: Unfortunately, OLM will not support an operator communicating its status via cluster-scoped CRs. If an operator is scoped to a namespace but writes conditions to a cluster-scoped CR, OLM would be unable to identify which operator is communicating its status if multiple installs of the operator were present on cluster.
```

Comment:
What if we include the name of the probe as another annotation applied to individual CRs to disambiguate which `Probe(s)` they are related to.

Taking this a little further, what if all the annotations were per-CR or resource, and putting condition annotations on a CRD was just syntax sugar which projected the annotations to all CRs on the cluster.

Orthogonal: I think we should also point out that we're excluding CRs provided by APIServices w/ the design as well.

Reply:
> What if we include the name of the probe as another annotation applied to individual CRs to disambiguate which Probe(s) they are related to.

If OLM is responsible for adding this probe annotation, I do not know how OLM could identify which CRs are owned by the operator at the cluster level. The operator would need to add these annotations itself and know about the Probe resource on cluster. At that point, the operator could probably integrate with OLM's probe resource directly.

> Taking this a little further, what if all the annotations were per-CR or resource, and putting condition annotations on a CRD was just syntax sugar which projected the annotations to all CRs on the cluster.

While this is more flexible, I believe it could introduce a lot of additional complexity to end users. If each and every resource CR can communicate readiness in a variety of different ways it seems like it would be far more difficult to understand the state of your cluster at any given point. If an operator author wanted such flexibility, it seems like they should integrate more tightly with OLM and just manage the probe resource themselves     .

Reply:
> If OLM is responsible for adding this probe annotation,

The operator author or operator would be responsible for adding this annotation. 

>  I do not know how OLM could identify which CRs are owned by the operator at the cluster level.

If you add the ability to annotate per-CR, you give users the ability to select the cluster-scoped CR that should be viewed.

> If each and every resource CR can communicate readiness in a variety of different ways it seems like it would be far more difficult to understand the state of your cluster at any given point. If an operator author wanted such flexibility, it seems like they should integrate more tightly with OLM and just manage the probe resource themselves .

but we already have this with the current design, just at the CRD level. I can define X distinct CRD condition annotations for an operator that provides X CRDs. In both situations, it's up to the `Probe` resource to reduce all of the conditions.